### ISSUE TITLE
Different results with pytorch and skorch

### ISSUE THREAD

**Issue by sliwy** [skorch-dev/skorch#591]
I really like the idea of skorch, so I am trying to replace my training loop with the skorch one. At this point, I can't figure out why I get different results using skorch and using my PyTorch training loop. From my perspective, all the hyperparameters are the same, but in my training scenario, I achieve training accuracy of 100% after around 20-30 epochs while in skorch I can't achieve it even after 100 epochs. I do not expect exactly the same results (enough for me to be similar).
Can you see any difference between skorch and my training loop (`train` function)? Am I missing something? You can find whole script here: https://github.com/sliwy/braindecode/blob/strange_example/examples/skorch_issue.py
Below fragment with my function:
```
def train(net, dataset_train, dataset_val, lr=0.001, batch_size=200, epochs=100,
          device='cpu'):

    dataloader_train = DataLoader(dataset_train, batch_size=batch_size,
                                  shuffle=True)
    dataloader_val = DataLoader(dataset_val, batch_size=batch_size,
                                shuffle=False)

    accuracy_epochs_train = []
    accuracy_epochs_val = []
    loss_epochs_train = []
    loss_epochs_val = []

    loss_fn = nn.NLLLoss()
    optimizer = torch.optim.Adam(net.parameters(), lr=lr)
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, epochs)

    try:
        net.train()
        for epoch in range(epochs):
            # TRAIN
            loss_one_epoch_train = []
            accuracy_one_epoch_train = []
            batches_len = []
            for batch in dataloader_train:
                x, y = batch
                batches_len.append(x.shape[0])
                y_pred = net(x.float().to(device))
                loss = loss_fn(y_pred, y.long().to(device))

                loss.backward()
                optimizer.step()
                optimizer.zero_grad()

                accuracy_one_epoch_train.append(accuracy_score(
                    y, np.argmax(y_pred.cpu().detach().numpy(), axis=1))
                )
                loss_one_epoch_train.append(loss.cpu().detach().numpy())

            scheduler.step()
            accuracy_epochs_train.append(np.average(accuracy_one_epoch_train,
                                                    weights=batches_len))
            loss_epochs_train.append(np.average(loss_one_epoch_train,
                                                weights=batches_len))

            # VALIDATION
            net.eval()
            loss_one_epoch_val = []
            accuracy_one_epoch_val = []
            batches_len = []
            with torch.no_grad():
                for batch in dataloader_val:
                    x, y = batch
                    batches_len.append(x.shape[0])
                    y_pred_val = net(x.float().to(device))
                    accuracy_one_epoch_val.append(accuracy_score(
                        y,
                        np.argmax(y_pred_val.cpu().detach().numpy(), axis=1)))
                    loss_one_epoch_val.append(loss_fn(
                        y_pred_val, y.long().to(device)).cpu().detach().numpy())

            accuracy_epochs_val.append(np.average(accuracy_one_epoch_val,
                                                  weights=batches_len))
            loss_epochs_val.append(np.average(loss_one_epoch_val,
                                              weights=batches_len))

            print('Epoch: ', epoch,
                  ' Train loss: {:.4f}'.format(loss_epochs_train[-1]),
                  ' Train accuracy {:.4f}'.format(accuracy_epochs_train[-1]),
                  ' Val loss: {:.4f}'.format(loss_epochs_val[-1]),
                  ' Val accuracy: {:.4f}'.format(accuracy_epochs_val[-1]))

    except KeyboardInterrupt:
        pass
    net.eval()
    return net, dict(accuracy_train=accuracy_epochs_train,
                     loss_train=loss_epochs_train,
                     accuracy_val=accuracy_epochs_val,
                     loss_val=loss_epochs_val)
```
and here skorch model:
```
clf = MyClassifier(
    model_skorch,
    criterion=torch.nn.NLLLoss,
    optimizer=torch.optim.Adam,
    optimizer__lr=lr,
    train_split=splitter,
    batch_size=batch_size,
    lr=lr,
    callbacks=[
        ('train_acc', EpochScoring(
            'accuracy', name='train_accuracy', on_train=True,
            lower_is_better=False)),
        ('valid_acc', EpochScoring(
            'accuracy', name='valid_acc', on_train=False,
            lower_is_better=False)),
        ('lr', LRScheduler(policy='CosineAnnealingLR', T_max=epochs))
    ],
    iterator_train=DataLoader,
    iterator_train__shuffle=True,
    iterator_valid=DataLoader,
    iterator_valid__shuffle=False,
    device=device,
    warm_start=False
)

train_results = train(model_old, ds_train, ds_valid, lr=lr,
                      batch_size=batch_size, epochs=epochs, device=device)
clf.fit(train_valid_set, y=None, epochs=epochs)
```
Output of trainings:
```
Epoch:  0  Train loss: 1.8632  Train accuracy 0.2391  Val loss: 1.3636  Val accuracy: 0.2759
Epoch:  1  Train loss: 1.3749  Train accuracy 0.2609  Val loss: 1.3710  Val accuracy: 0.3793
Epoch:  2  Train loss: 1.3370  Train accuracy 0.4130  Val loss: 1.3835  Val accuracy: 0.3793
Epoch:  3  Train loss: 1.3181  Train accuracy 0.4174  Val loss: 1.3005  Val accuracy: 0.5000
Epoch:  4  Train loss: 1.2385  Train accuracy 0.5652  Val loss: 1.2897  Val accuracy: 0.2931
Epoch:  5  Train loss: 1.2092  Train accuracy 0.4391  Val loss: 1.2521  Val accuracy: 0.3448
Epoch:  6  Train loss: 1.1138  Train accuracy 0.5652  Val loss: 1.2043  Val accuracy: 0.4310
Epoch:  7  Train loss: 1.0379  Train accuracy 0.5391  Val loss: 1.1714  Val accuracy: 0.4655
Epoch:  8  Train loss: 0.9798  Train accuracy 0.5217  Val loss: 1.1031  Val accuracy: 0.4828
Epoch:  9  Train loss: 0.8597  Train accuracy 0.7478  Val loss: 0.9977  Val accuracy: 0.6034
Epoch:  10  Train loss: 0.7804  Train accuracy 0.7913  Val loss: 0.9320  Val accuracy: 0.5862
Epoch:  11  Train loss: 0.7075  Train accuracy 0.7913  Val loss: 0.9682  Val accuracy: 0.5000
Epoch:  12  Train loss: 0.6186  Train accuracy 0.8043  Val loss: 0.9335  Val accuracy: 0.5172
Epoch:  13  Train loss: 0.5687  Train accuracy 0.7957  Val loss: 0.9558  Val accuracy: 0.5000
Epoch:  14  Train loss: 0.4932  Train accuracy 0.8522  Val loss: 0.8607  Val accuracy: 0.5172
Epoch:  15  Train loss: 0.4341  Train accuracy 0.8826  Val loss: 0.8764  Val accuracy: 0.5862
Epoch:  16  Train loss: 0.3761  Train accuracy 0.9000  Val loss: 0.8347  Val accuracy: 0.5862
Epoch:  17  Train loss: 0.3515  Train accuracy 0.9174  Val loss: 1.2631  Val accuracy: 0.4138
Epoch:  18  Train loss: 0.3948  Train accuracy 0.8565  Val loss: 0.8684  Val accuracy: 0.5517
Epoch:  19  Train loss: 0.2275  Train accuracy 0.9739  Val loss: 0.8961  Val accuracy: 0.5345
Epoch:  20  Train loss: 0.1846  Train accuracy 0.9826  Val loss: 0.8780  Val accuracy: 0.5172
Epoch:  21  Train loss: 0.1536  Train accuracy 0.9957  Val loss: 0.8497  Val accuracy: 0.6034
Epoch:  22  Train loss: 0.1492  Train accuracy 0.9957  Val loss: 1.0204  Val accuracy: 0.4655
Epoch:  23  Train loss: 0.1383  Train accuracy 0.9870  Val loss: 0.9146  Val accuracy: 0.6207
Epoch:  24  Train loss: 0.0959  Train accuracy 1.0000  Val loss: 0.9546  Val accuracy: 0.5862
Epoch:  25  Train loss: 0.1038  Train accuracy 0.9957  Val loss: 0.9902  Val accuracy: 0.5345
Epoch:  26  Train loss: 0.0737  Train accuracy 0.9957  Val loss: 1.0328  Val accuracy: 0.4828
Epoch:  27  Train loss: 0.0646  Train accuracy 1.0000  Val loss: 0.9180  Val accuracy: 0.6207
Epoch:  28  Train loss: 0.0551  Train accuracy 1.0000  Val loss: 0.9380  Val accuracy: 0.5517
Epoch:  29  Train loss: 0.0587  Train accuracy 1.0000  Val loss: 0.9312  Val accuracy: 0.6034
Epoch:  30  Train loss: 0.0433  Train accuracy 1.0000  Val loss: 1.0237  Val accuracy: 0.5172
Epoch:  31  Train loss: 0.0430  Train accuracy 1.0000  Val loss: 1.0625  Val accuracy: 0.4828
Epoch:  32  Train loss: 0.0406  Train accuracy 1.0000  Val loss: 1.0198  Val accuracy: 0.5690
Epoch:  33  Train loss: 0.0334  Train accuracy 1.0000  Val loss: 1.0027  Val accuracy: 0.6034
Epoch:  34  Train loss: 0.0324  Train accuracy 1.0000  Val loss: 1.0075  Val accuracy: 0.5690
Epoch:  35  Train loss: 0.0319  Train accuracy 1.0000  Val loss: 1.0013  Val accuracy: 0.5862
Epoch:  36  Train loss: 0.0290  Train accuracy 1.0000  Val loss: 1.0016  Val accuracy: 0.6034
Epoch:  37  Train loss: 0.0269  Train accuracy 1.0000  Val loss: 1.0096  Val accuracy: 0.5690
Epoch:  38  Train loss: 0.0263  Train accuracy 1.0000  Val loss: 1.0131  Val accuracy: 0.5517
Epoch:  39  Train loss: 0.0258  Train accuracy 1.0000  Val loss: 1.0092  Val accuracy: 0.5690
Epoch:  40  Train loss: 0.0251  Train accuracy 1.0000  Val loss: 1.0041  Val accuracy: 0.6207
Epoch:  41  Train loss: 0.0243  Train accuracy 1.0000  Val loss: 1.0008  Val accuracy: 0.6034
Epoch:  42  Train loss: 0.0239  Train accuracy 1.0000  Val loss: 0.9997  Val accuracy: 0.6034
Epoch:  43  Train loss: 0.0235  Train accuracy 1.0000  Val loss: 0.9993  Val accuracy: 0.6034
Epoch:  44  Train loss: 0.0233  Train accuracy 1.0000  Val loss: 0.9989  Val accuracy: 0.6034
Epoch:  45  Train loss: 0.0232  Train accuracy 1.0000  Val loss: 0.9991  Val accuracy: 0.6034
Epoch:  46  Train loss: 0.0231  Train accuracy 1.0000  Val loss: 0.9994  Val accuracy: 0.6034
Epoch:  47  Train loss: 0.0231  Train accuracy 1.0000  Val loss: 0.9996  Val accuracy: 0.6034
Epoch:  48  Train loss: 0.0230  Train accuracy 1.0000  Val loss: 0.9998  Val accuracy: 0.6034
Epoch:  49  Train loss: 0.0230  Train accuracy 1.0000  Val loss: 0.9998  Val accuracy: 0.6034
  epoch    train_accuracy    train_loss    valid_acc    valid_loss     dur
-------  ----------------  ------------  -----------  ------------  ------
      1            0.2739        1.9750       0.2586        1.3937  4.3725
      2            0.2913        1.8846       0.2586        1.3973  4.3467
      3            0.2870        1.7209       0.2414        1.5438  5.0067
      4            0.3435        1.6073       0.2069        1.7565  4.2775
      5            0.2957        1.6534       0.1897        1.9812  4.3362
      6            0.3043        1.6340       0.2414        2.2584  4.4270
      7            0.2826        1.7242       0.2759        2.6736  4.3656
      8            0.3609        1.5921       0.2069        3.1249  4.3162
      9            0.3304        1.6719       0.2241        3.4644  4.3399
     10            0.2913        1.6622       0.2241        3.6452  4.2021
     11            0.4087        1.4728       0.2241        3.7608  5.3019
     12            0.3913        1.4910       0.2241        3.8962  5.4395
     13            0.4435        1.3117       0.2241        4.0305  4.5665
     14            0.4174        1.4031       0.2241        4.1754  4.9790
     15            0.4826        1.2982       0.2241        4.3644  4.6259
     16            0.4435        1.2174       0.2241        4.5012  4.1458
     17            0.4696        1.3127       0.2241        4.5560  4.7421
     18            0.4609        1.2107       0.2241        4.5897  4.7000
     19            0.4957        1.1653       0.2241        4.5925  4.3273
     20            0.4826        1.1375       0.2241        4.5741  4.1822
     21            0.5435        1.1385       0.2241        4.6070  4.3922
     22            0.5217        1.1084       0.2241        4.6664  4.3093
     23            0.5435        1.0387       0.2241        4.7515  4.7746
     24            0.5609        1.0467       0.2241        4.8036  4.7014
     25            0.5739        0.9471       0.2241        4.8487  4.7985
     26            0.5435        1.0380       0.2241        4.9046  5.5395
     27            0.5304        1.0324       0.2241        4.8726  5.2462
     28            0.5913        0.9871       0.2241        4.8970  4.3620
     29            0.5565        0.9935       0.2241        4.9289  4.4241
     30            0.5609        0.9968       0.2241        5.0773  4.1917
     31            0.5957        0.9779       0.2241        5.2237  4.7678
     32            0.5652        0.9738       0.2241        5.4271  4.6018
     33            0.5739        0.9524       0.2241        5.5707  4.2368
     34            0.6043        0.9721       0.2241        5.6523  4.2515
     35            0.5870        0.9149       0.2241        5.7230  4.4908
     36            0.5826        1.0017       0.2241        5.7470  3.8612
     37            0.5957        1.0200       0.2241        5.8070  5.4109
     38            0.6087        0.9449       0.2241        5.8637  6.5441
     39            0.5957        0.9576       0.2241        5.9983  5.4039
     40            0.5696        0.9888       0.2241        5.9859  6.2505
     41            0.6478        0.8245       0.2241        6.0316  5.2166
     42            0.5913        0.9557       0.2241        6.0487  5.0322
     43            0.6087        0.9408       0.2241        6.1020  5.5927
     44            0.5870        0.8885       0.2241        6.1553  4.7618
     45            0.6087        0.9088       0.2241        6.1860  5.1060
     46            0.6304        0.8764       0.2241        6.2038  5.8812
     47            0.6304        0.9121       0.2241        6.2040  5.7731
     48            0.6261        0.8348       0.2241        6.2600  5.5920
     49            0.6304        0.8479       0.2241        6.2570  5.6965
     50            0.6130        0.9659       0.2241        6.2175  5.3494

```

In case you would like to reproduce it, please use the script from link above. Before running you need to install skorch, download and install braindecode package (with model definition, I didn't want to copy-paste too much code) from https://github.com/braindecode/braindecode.git and download around 25Mb of data used for this training https://drive.google.com/open?id=1szaHV-mroHenOTKruY2ecpVQq8ZN9ENF (place it together with the script).

**Comment by BenjaminBossan** [comment]
I didn't try out your code but have some questions:

* Did you make sure the train/valid split is exactly the same?
* Could you please post the exact module you used for your own test and for skorch? (link to the code is sufficient)
* Could you try to remove some of the complexity, e.g. no train/valid split, no callbacks, and see if you get the same results then?

**Comment by sliwy** [comment]
@BenjaminBossan thank you for your reply. Your questions were very helpful, as I simplified the problem I found that adding validation to my function changes the training process and then I've found that  I am setting my model into `train` mode only once at the beginning and after first epoch validation, the model was for the rest of training in the `eval` mode, so it affected the results. Thanks for the help.
